feat: add Next.js server-side auth and HTTP-only cookie content (#8224)
aws-amplify/docs: src/pages/[platform]/build-a-backend/server-side-rendering/index.mdx

`ConfigureAmplifyClientSide.ts`:

> ```typescript title="src/components/ConfigureAmplifyClientSide.tsx"

Reviewer: ````suggestion
```tsx title="src/components/ConfigureAmplifyClientSide.tsx"
````

`layout.tsx`:

> ```jsx title="src/app/layout.tsx"

Reviewer: ````suggestion
```tsx title="src/app/layout.tsx"
````

| `/api/auth/sign-in?provider=<social-provider-name>` | Upon navigating an end user to this route, they’ll be redirected first to the Amazon Cognito Managed Login and then the specified social provider sign-in page. After sign-in, they’ll be redirected back to the route `/api/auth/sign-in-callback`. |
| `/api/auth/sign-out`                                | Upon navigating an end user to this route, the end user will be signed out and redirected to the route `/api/auth/sign-out-callback`. |
| `/api/auth/sign-in-callback`                        | Amazon Cognito Managed Login redirects an end user back to this route after signing in. Amplify exchanges auth tokens and stores them as HttpOnly cookies in the browser cookie store, then redirects the end user back to the route specified by the `redirectOnSignInComplete` parameter. |
| `/api/auth/sign-out-callback`                       | Amazon Cognito Managed Login redirects an end user back to this route after signing out, Amplify revokes access token and refresh token and removes token cookies from browser cookie store, then redirects the end user back to the route specified by the `redirectOnSignOutComplete` parameter. |

> > **Note:** A signing-out call involves multiple steps, including signing out from Amazon Cognito Managed Login, revoking tokens, and removing cookies. If the user closes the browser during the process, the following may occur:

Reviewer: can we pivot this to a Callout?

Author: Updated.

>
> 1. auth token have not been revoked - user remains signed in
> 2. auth token have been revoked but cookies have not been removed - cookies will be removed when the user visits the app again

#### Step 4 - Provide the redirect URLs to the Auth Resource in Amplify

You can provide the callback API routes as the redirect URLs in the Auth resource configuration. For example:

```ts title="amplify/auth/resource.ts"
export const auth = defineAuth({
  loginWith: {
    email: true,
    // highlight-start
    externalProviders: {
      callbackUrls: ["https://myapp.com/api/auth/sign-in-callback"],
      logoutUrls: ["https://myapp.com/api/auth/sign-out-callback"],
    },
    // highlight-end
  },
});
```

This enables Amazon Cognito Hosted UI to support the server-side authentication flows. You may upgrade to the latest Amazon Cognito Managed Login Branding to customize the sign-in and sign-up pages. See [Amazon Cognito user pool managed login](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-managed-login.html) for more information.

> #### Step 5 - Use Anchor link for initiating server-side authentication flows

Reviewer 2: just minor nit: I think the example here could be just one page server side on `page.tsx` example. Rather than these button components. 

Since i think these components are just given and not told where to use it. And new to next devs can put it on client side as well.

Author: Unfortunately using anchor link directly in page.tsx trigger Next eslint rules, and it recommends to use `Link` component provided by Next.js, which enforce to prefetch linked resource and cause unexpected errors. So we give button example to ensure it works out of box.

</Callout>

### (Preview) Perform authentication on the server side and enable HTTP-only cookies

<Callout warning>

**NOTE:** Once you enable the server-side sign-in feature, auth tokens are stored in HTTP-only cookies and you may not change the HTTP-only attribute. Since these cookies are inaccessible from client-side scripts, you won’t be able to use any Amplify JS APIs on the client side. Therefore, you don’t need to configure Amplify on the client side.

</Callout>

**Prerequisites**

To authenticate users on the server side, you must enable either Amazon Cognito Managed Login or Hosted UI in your Amazon Cognito User Pool client.

**Step 1: Specify the origin of your app in environment variables**

Add the following environment variables to your Next.js app. For example in a `.env` file:

```shell title=".env"
AMPLIFY_APP_ORIGIN=https://myapp.com
```

Ensure this environment variables is accessible in your Next.js app's server runtime.

**Step 2 - Export the `createAuthRouteHandlers` function**

> `createAuthRouteHandlers` function is created by the `createServerRunner` function call when you configure Amplify for server-side usage. You can export this function from your `amplifyServerUtils.ts` file. You can also configure cookie attributes with the `runtimeOptions` parameter.

Author: Yes the parameter name is still `runtimeOptions`.

```typescript title="utils/amplifyServerUtils.ts"
import { createServerRunner } from '@aws-amplify/adapter-nextjs';

> import config from '@/amplifyconfiguration.json';

Author: Good catch

You can provide the callback API routes as the redirect URLs in the Auth resource configuration. For example:

```ts title="amplify/auth/resource.ts"
export const auth = defineAuth({
  loginWith: {
    email: true,
    externalProviders: {
      google: {...},
      // highlight-start
      callbackUrls: ["https://myapp.com/api/auth/sign-in-callback"],
      logoutUrls: ["https://myapp.com/api/auth/sign-out-callback"],
      // highlight-end
    },
  },
});
```
**Step 5 - Use Anchor link for initiating server-side authentication flows**

Use HTML anchor links to navigate users to the sign-in and sign-up routes. For example:

<BlockSwitcher>
<Block name="Sign in button">
```tsx title="src/components/SignInButton.tsx"
export const SignInButton() {
  return (
    <a href="/api/auth/sign-in">
      Sign In
    </a>
  );
}
```
</Block>
<Block name="Sign in with Google button">
```tsx title="src/components/SignInWithGoogleButton.tsx"
export const SignInButton() {
  return (
    <a href="/api/auth/sign-in?provider=Google">
      Sign In
    </a>
  );
}
```
</Block>
<Block name="Sign up button">
```tsx title="src/components/SignUpButton.tsx"

> export const SignInButton() {

Reviewer 3: ```suggestion
export const SignUpButton() {
```

  return (
    <a href="/api/auth/sign-up">

>       Sign In

Reviewer 3: ```suggestion
      Sign Up
```

    </a>
  );
}
```
</Block>
<Block name="Sign out button">
```tsx title="src/components/SignOutButton.tsx"

> export const SignInButton() {

Reviewer 3: ```suggestion
export const SignOutButton() {
```

  return (
    <a href="/api/auth/sign-out">

>       Sign In

Reviewer 3: ```suggestion
      Sign Out
```

</Callout>

> ### (Preview) Perform authentication on the server side and enable HTTP-only cookies

Reviewer: ```suggestion
### (Preview) Perform authentication on the server side and enable HttpOnly cookies
```

nit for the attribute name

Author: Updated.

<Callout warning>

**NOTE:** Once you enable the server-side sign-in feature, auth tokens are stored in HTTP-only cookies and you may not change the HTTP-only attribute. Since these cookies are inaccessible from client-side scripts, you won’t be able to use any Amplify JS APIs on the client side. Therefore, you don’t need to configure Amplify on the client side.

</Callout>

**Prerequisites**

To authenticate users on the server side, you must enable either Amazon Cognito Managed Login or Hosted UI in your Amazon Cognito User Pool client.

> **Step 1 - Specify the origin of your app in environment variables**

Reviewer: can we convert the steps to h4's?

Author: updated.

Add the following environment variables to your Next.js app. For example in a `.env` file:

> ```shell title=".env"

Reviewer: ````suggestion
```shell title=".env" showLineNumbers={false}
````

does this need to be added to the Hosting env as well?

AMPLIFY_APP_ORIGIN=https://myapp.com
```

Ensure this environment variables is accessible in your Next.js app's server runtime.

> **Note:** Token cookies are transmitted via server-side authentication flows. In production environments, it is recommended to use HTTPS as the origin for enhanced security.

**Step 2 - Export the `createAuthRouteHandlers` function**

The `createAuthRouteHandlers` function is created by the `createServerRunner` function call when you configure Amplify for server-side usage. You can export this function from your `amplifyServerUtils.ts` file. You can also configure cookie attributes with the `runtimeOptions` parameter.

```typescript title="utils/amplifyServerUtils.ts"
import { createServerRunner } from '@aws-amplify/adapter-nextjs';
import config from '@/amplify_outputs.json';

export const {

>   runWithAmplifyServerContext

Reviewer 2: ```suggestion
  runWithAmplifyServerContext,
```

> import config from '@/amplify_outputs.json';
> 
> export const {
>   runWithAmplifyServerContext
>   // highlight-start
>   createAuthRouteHandlers,
>   // highlight-end
> } = createServerRunner({
>   config,

Reviewer: ```suggestion
import outputs from '@/amplify_outputs.json';

export const {
  runWithAmplifyServerContext
  // highlight-start
  createAuthRouteHandlers,
  // highlight-end
} = createServerRunner({
  config: outputs,
```

Author: updated.

  // highlight-start
  runtimeOptions: {
    cookies: {
      domain: '.myapp.com', // making cookies available to all subdomains
      sameSite: 'strict',
      maxAge: 60 * 60 * 24 * 7 // 7 days
    }
  }
  // highlight-end
});
```

**Step 3 - Set up the Auth API routes**

Create an API route using the `createAuthRouteHandlers` function. For example:

<BlockSwitcher>
<Block name="App router">
```typescript title="src/app/api/auth/[slug].ts"

> import { createAuthRouteHandlers } from "@/amplifyServerUtils";

Reviewer 2: ```suggestion
import { createAuthRouteHandlers } from "@/utils/amplifyServerUtils";
```

Author: Good catch! updated.

export const GET = createAuthRouteHandlers({
  redirectOnSignInComplete: "/home",
  redirectOnSignOutComplete: "/sign-in",
});
```
</Block>
<Block name="Pages router">
```typescript title="src/pages/api/auth/[slug].ts"

> import { createAuthRouteHandlers } from "@/amplifyServerUtils";

Reviewer 2: ```suggestion
import { createAuthRouteHandlers } from "@/utils/amplifyServerUtils";
```

| `/api/auth/sign-out-callback`                       | Amazon Cognito Managed Login redirects an end user back to this route after signing out, Amplify revokes access token and refresh token and removes token cookies from browser cookie store, then redirects the end user back to the route specified by the `redirectOnSignOutComplete` parameter. |

> **Note:** A signing-out call involves multiple steps, including signing out from Amazon Cognito Managed Login, revoking tokens, and removing cookies. If the user closes the browser during the process, the following may occur:
>
> 1. auth token have not been revoked - user remains signed in
> 2. auth token have been revoked but cookies have not been removed - cookies will be removed when the user visits the app again

**Step 4 - Provide the redirect URLs to the Auth Resource in Amplify**

You can provide the callback API routes as the redirect URLs in the Auth resource configuration. For example:

```ts title="amplify/auth/resource.ts"
export const auth = defineAuth({
  loginWith: {
    email: true,
    externalProviders: {

>       google: {...},

Reviewer: ```suggestion
      google: {/* ... */},
```

</Callout>

### (Experimental) Perform authentication on the server side and enable HttpOnly cookies

<Callout warning>

> **NOTE:** Once you enable the server-side sign-in feature, auth tokens are stored in HttpOnly cookies and you may not change the HttpOnly attribute. Since these cookies are inaccessible from client-side scripts, you won’t be able to use any Amplify JS APIs on the client side. Therefore, you don’t need to configure Amplify on the client side. You can keep using [these Amplify JS server-side APIs](/[platform]/build-a-backend/server-side-rendering/#supported-apis-for-nextjs-server-side-usage) on the server side.

Reviewer: ```suggestion
**Warning:** Once you enable the server-side sign-in feature, auth tokens are stored in HttpOnly cookies and you may not change the HttpOnly attribute. Since these cookies are inaccessible from client-side scripts, you won’t be able to use any Amplify JS APIs on the client side. Therefore, you don’t need to configure Amplify on the client side. You can keep using [these Amplify JS server-side APIs](/[platform]/build-a-backend/server-side-rendering/#supported-apis-for-nextjs-server-side-usage) on the server side.
```

nit for warning callout to use "Warning" text, align casing with other notes

</Callout>

**Prerequisites**

To authenticate users on the server side, you must enable either Amazon Cognito Managed Login or Hosted UI in your Amazon Cognito User Pool client.

#### Step 1 - Specify the origin of your app in environment variables

Add the following environment variables to your Next.js app. For example in a `.env` file:

```shell title=".env" showLineNumbers={false}
AMPLIFY_APP_ORIGIN=https://myapp.com
```

> Ensure this environment variables is accessible in your Next.js app's server runtime.

Reviewer 4: ```suggestion
Ensure this environment variable is accessible in your Next.js app's server runtime.
```

> > **Note:** Token cookies are transmitted via server-side authentication flows. In production environments, it is recommended to use HTTPS as the origin for enhanced security.

Reviewer: can we use a callouts for these notes instead of the blockquotes?
```suggestion
<Callout info>

**Note:** Token cookies are transmitted via server-side authentication flows. In production environments, it is recommended to use HTTPS as the origin for enhanced security.

</Callout>
```

#### Step 2 - Export the `createAuthRouteHandlers` function

The `createAuthRouteHandlers` function is created by the `createServerRunner` function call when you configure Amplify for server-side usage. You can export this function from your `amplifyServerUtils.ts` file. You can also configure cookie attributes with the `runtimeOptions` parameter.

```typescript title="src/utils/amplifyServerUtils.ts"
import { createServerRunner } from '@aws-amplify/adapter-nextjs';
import outputs from '@/amplify_outputs.json';

export const {
  runWithAmplifyServerContext,
  // highlight-start
  createAuthRouteHandlers,
  // highlight-end
} = createServerRunner({
  config: outputs,
  // highlight-start
  runtimeOptions: {
    cookies: {
      domain: '.myapp.com', // making cookies available to all subdomains
      sameSite: 'strict',
      maxAge: 60 * 60 * 24 * 7 // 7 days
    }
  }
  // highlight-end
});
```

#### Step 3 - Set up the Auth API routes

Create an API route using the `createAuthRouteHandlers` function. For example:

<BlockSwitcher>
<Block name="App router">

> ```typescript title="src/app/api/auth/[slug].ts"

Reviewer 4: ```suggestion
```typescript title="src/app/api/auth/[slug]/route.ts"
```

Author: Good catch ...

  redirectOnSignOutComplete: "/sign-in",
});
```
</Block>
<Block name="Pages router">
```typescript title="src/pages/api/auth/[slug].ts"
import { createAuthRouteHandlers } from "@/utils/amplifyServerUtils";

export default createAuthRouteHandlers({
  redirectOnSignInComplete: "/home",
  redirectOnSignOutComplete: "/sign-in",
});
```
</Block>
</BlockSwitcher>

With the above example, Amplify generates the following API routes:

| API Routes                                          | What it does                                                 |
| --------------------------------------------------- | ------------------------------------------------------------ |
| `/api/auth/sign-up`                                 | Upon navigating an end user to this route, they’ll be redirected to the Amazon Cognito Managed Login sign-up form. After sign-up and sign-in, they’ll be redirected back to the route `/api/auth/sign-in-callback`. |
| `/api/auth/sign-in`                                 | Upon navigating an end user to this route, they’ll be redirected to the Amazon Cognito Managed Login sign-in form. After sign-in, they’ll be redirected back to the route `/api/auth/sign-in-callback`. |

> | `/api/auth/sign-in?provider=<social-provider-name>` | Upon navigating an end user to this route, they’ll be redirected to first to the Amazon Cognito Managed Login and then the specified social provider sign-in page. After sign-in, they’ll be redirected back to the route `/api/auth/sign-in-callback`. |

Reviewer 4: ```suggestion
| `/api/auth/sign-in?provider=<social-provider-name>` | Upon navigating an end user to this route, they’ll be redirected first to the Amazon Cognito Managed Login and then the specified social provider sign-in page. After sign-in, they’ll be redirected back to the route `/api/auth/sign-in-callback`. |
```

</Callout>

> ### (Experimental) Perform authentication on the server side and enable HttpOnly cookies

Reviewer 2: As working through testnig, I found this quiet confusing and interfering with current flow. Could this be a separate page in itself?
